examples_tests/35.TAA/taa.frag

```diff
@@ -0,0 +1,154 @@
+	float ray_len = chebyshevNorm(hist_);
+	if (ray_len > 1.0)
+	{
+		// calculate intersection of unit sphere with ray center->hist_ (i.e. normalize)
+		vec3 inter1 = hist_/ray_len;
+		// scale back to get intersection with ellipsoid and translate to original space
+		return inter1*scale + center;
+	}
+	else
+		return hist; // hist sample is within neighbourhood color space
+}
+
+//finds texel offset (relative to `uv`) of depth-wise closest fragment in 3x3 neighbourhood
+ivec2 closestFragmentOffsetIn3x3(in vec2 uv)
+{
+	vec3 d0 = vec3(-1.0, 1.0, textureLodOffset(uDepthBuf, uv, 0.0, ivec2(-1, 1)).x);
+	vec3 d1 = vec3(0.0, 1.0, textureLodOffset(uDepthBuf, uv, 0.0, ivec2(0, 1)).x);
+	vec3 d2 = vec3(1.0, 1.0, textureLodOffset(uDepthBuf, uv, 0.0, ivec2(1, 1)).x);
+	vec3 d3 = vec3(-1.0, 0.0, textureLodOffset(uDepthBuf, uv, 0.0, ivec2(-1, 0)).x);
+	vec3 d4 = vec3(0.0, 0.0, textureLod(uDepthBuf, uv, 0.0).x);
+	vec3 d5 = vec3(1.0, 0.0, textureLodOffset(uDepthBuf, uv, 0.0, ivec2(1, 0)).x);
+	vec3 d6 = vec3(-1.0, -1.0, textureLodOffset(uDepthBuf, uv, 0.0, ivec2(-1, -1)).x);
+	vec3 d7 = vec3(0.0, -1.0, textureLodOffset(uDepthBuf, uv, 0.0, ivec2(0, -1)).x);
+	vec3 d8 = vec3(1.0, -1.0, textureLodOffset(uDepthBuf, uv, 0.0, ivec2(1, -1)).x);
+	
+	//less = further from screen
+#define _CLOSEST(a, b) mix(a, b, a.z>b.z)
+	vec3 closest = d0;
+	closest = _CLOSEST(closest, d1);
+	closest = _CLOSEST(closest, d2);
+	closest = _CLOSEST(closest, d3);
+	closest = _CLOSEST(closest, d4);
+	closest = _CLOSEST(closest, d5);
+	closest = _CLOSEST(closest, d6);
+	closest = _CLOSEST(closest, d7);
+	closest = _CLOSEST(closest, d8);
+#undef _CLOSEST
+	
+	return ivec2(closest.xy);
+}
+
+void main()
+{
+/*
+	float depth = texture(uDepthBuf, TexCoords).x; // linearize depth?
+	
+	vec4 clip = vec4(2.0*vec3(TexCoords, depth) - 1.0, 1.0);
+	vec4 world = invVP*clip;
+	world = world/world.w;
+	
+	vec4 q_cs = uPrevVP*world;
+	vec2 q_uv = 0.5*(q_cs.xy/q_cs.w) + 0.5;
+	vec2 vel = TexCoords - q_uv;
+*/
+	// TODO test later whether closestFragmentOffsetIn3x3() is worth it
+	vec2 vel = textureLodOffset(uVelocityBuf, TexCoords, 0.0, closestFragmentOffsetIn3x3(TexCoords)).xy;
+	
+	vec2 q_uv = TexCoords - vel;
+	
+	vec3 current = textureLod(uCurrentBuf, TexCoords, 0.0).rgb;
+	vec3 history = textureLod(uHistoryBuf, q_uv, 0.0).rgb;
+	vec3 neigh0 = textureLodOffset(uCurrentBuf, TexCoords, 0.0, ivec2(1, 0)).rgb;
+	vec3 neigh1 = textureLodOffset(uCurrentBuf, TexCoords, 0.0, ivec2(0, 1)).rgb;
+	vec3 neigh2 = textureLodOffset(uCurrentBuf, TexCoords, 0.0, ivec2(-1, 0)).rgb;
+	vec3 neigh3 = textureLodOffset(uCurrentBuf, TexCoords, 0.0, ivec2(0, -1)).rgb;
+
+	vec3 aabb_min = min(current, min(neigh0, min(neigh1, min(neigh2, neigh3))));
+	vec3 aabb_max = max(current, max(neigh0, max(neigh1, max(neigh2, neigh3))));
+
+	history = constrainHistSample(history, aabb_min, aabb_max);
+	
+	vec3 current_unjittered = textureLod(uCurrentBuf, TexCoords-uJitter, 0.0).rgb;
```

Comment:
see Brian Karis' 2013 TemporalAA presentation where he mentions Burely07 that Box filter does not behave well with moving samples, so for this tap (and this tap only) a pseudo-small-kernel-gaussian 2 wide pixel support reconstruction filter (save bandwidth with shared memory or approximating gaussian with 4=2x2 `textureLodOffset`).

Reply:
Filmic Temporal Supersample AA presentation mentions using a Bicubic fast appriximation (catmul-rom 5 tap bilinear thing)